### ISSUE TITLE
UMAP memory parallelisation issues

### ISSUE THREAD

**Issue by maikejulie** [lmcinnes/umap#429]
Hi UMAP team,

First off, thanks for the great work! I am trying to use UMAP to reduce the dimensionality of a dataset, where initial tests have shown very promising results. I’ve run into several issues (parallelization, memory management..) where some help would be great.

Context:
I’m working on a node in a cluster with ca 270GB available memory and 16 cores. The dataset I am currently working on is (3069899, 5), see below for example snippet. I would like to scale that up 12 times if possible.
Running UMAP in serial works for the (3069899, 5) dataset using only two dims and setting the random seed, but takes hours. Ideally I would like to run parameter sweeps to ensure meaningful results and reproducibility.

Issues:
- Removing the random seed doesn’t make it use more cores.
- Following https://github.com/lmcinnes/umap/issues/379, we tried installing pynndescent. The script seemed to use multiple cores on one of the nodes, after setting n_jobs=-1 in umap. But, due to memory issues, it did not run to completion. 
      -Example errors while invoking UMAP from within a DASK cluster are as follows:
            - distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 264.16 GB -- Worker memory limit: 270.51 GB
            - distributed.core - INFO - Event loop was unresponsive in Worker for 14.13s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
      - When the pandas dataframe was much smaller than the original, the calls to UMAP seem to run to completion (till saving results to a numpy array). 
      -low_memory=False in a couple of recent tests continued to throw memory exceptions.
- Dealing with a pending ‘small memory’, it has not worked testing on smaller machines.
- Calling UMAP defined inside a function to use Sciris to parallelise serial runs the following crop up:
      - It doesn’t save the results if called inside a function (see below). I tried np.save, pickle, turning it into an array again etc.. no luck.
      -It eats all available memory.

I have occasionally also (trying out e.g. different numba versions and whatnot) run into segfaults and ‘BUS error’ I believe are memory related.

My initial results were hugely promising, so I would very much appreciate any help you can offer!

Example code snippet that works: 

clusterable_embedding = umap.UMAP(
    n_neighbors=100,
    min_dist=0.0,
    n_components=2,
    random_state=48).fit_transform(pd_BV)

np.save('umap_data.npy', clusterable_embedding)

Defined in a function/save issue:

def runUMAP(neighbours, run):
    print(neighbours)
    clusterable_embedding = umap.UMAP(
        n_neighbors=neighbours,
        min_dist=0.0,
        n_components=2
    ).fit_transform(pd_BV)
    return clusterable_embedding

if __name__ == "__main__":

    neighbours = [60,80,120,150,180, 200,250]
    run = np.arange(1,5)
    [aa, bb] = np.meshgrid(neighbours,run)

    iterargs = list(zip(aa.flatten(), bb.flatten()))
    results=sc.parallelize(runUMAP, iterargs, ncpus=8)
    print('done')

    for ix, result in enumerate(results):
        np.save(f'save_{ix}.npy', result)

**Comment by lmcinnes** [comment]
@maikejulie thanks for the report, and sorry for the slow reply. This is all a little disconcerting. In terms of parallelism in UMAP, it is all handled via numba which creates a thread pool and handles all the thread dispatching itself. This means I don't have too much control (but it was a lot easier to get working). In practice the parallelism is only going to work on an SMP machine, so if you have a multi-node cluster it will not extend across nodes. Making a distributed version of both pynndescent and umap is something I would like to do, but it is rather non-trivial an dI don't expect to complete it any time soon; it would specifically involve brining dask into the internals.

The other catches are that umap (and pynndescent) can be pretty memory hungry. The numbers you report here do seem excessive to me, but it is not clear why that is happening. I have successfully run experimental runs of UMAP on 100,000,000 x 5,000,000 (sparse) datasets (admittedly on a large SMP server with 1TB of RAM), so your size of dataset should definitely be possible, but perhaps not on the hardware you have access to?

As for the models not saving -- I admit to being very confused by that. I have no experience with sciris, so I can't speak to what may be going on there. 

Sorry I don't have any better advice. Realistically I think the best option is to run the jobs serially. If you can get access to a large SMP node (many cores all using the same memory address space) then the parallelism internal to umap via numba should work there. Unfortunately it won't work in a multi-node environment with distributed memory.
